python/ray/data/_internal/plan.py

```diff
@@ -260,10 +260,13 @@ def schema(
         blocks = self._snapshot_blocks
         if not blocks:
             return None
-        # Don't force fetching in case it's a lazy block list, in which case we
-        # don't want to trigger full execution for a schema read. If we want to
-        # trigger execution to get schema, we'll trigger read tasks progressively
-        # until a viable schema is available, below.
+
+        # Only trigger the execution of first block in case it's a lazy block list.
+        # Don't trigger full execution for a schema read.
+        if isinstance(blocks, LazyBlockList):
+            blocks.compute_first_block()
+            blocks.ensure_metadata_for_first_block()
```

Comment:
This will mutate the snapshot blocks, so the computed first block will be cached for future executions, right?

Reply:
@clarkzinzow - yes the computed first block is cached. Hopefully these unit tests should cover the cases - https://github.com/ray-project/ray/blob/master/python/ray/data/tests/test_dataset.py#L1495-L1517.